File: matiasmasso/Mat.NET/DTO/DTOMenu.cs
using System;
using System.Collections.Generic;

namespace DTO
{
    public class DTOMenu : DTOBaseGuid
    {
        public DTOApp.AppTypes App { get; set; }
        public Cods Cod { get; set; }

        public DTOLangText Caption { get; set; }
        public DTOLangText LangUrl { get; set; }
        public string Ord { get; set; }
        public string Action { get; set; }
        public string Controller { get; set; }
        public List<DTORol> Rols { get; set; }
        public DTOMenu Parent { get; set; }
        public List<DTOMenu> Items { get; set; }

        public enum Cods
        {
            NotSet,
            Queries,
            Forms,
            Profile,
            Customer,
            Rep,
            Comercial,
            Staff,
            Product
        }

        public DTOMenu() : base()
        {
            Caption = new DTOLangText(base.Guid, DTOLangText.Srcs.WebMenuItem);
            Rols = new List<DTORol>();
            Items = new List<DTOMenu>();
        }

        public DTOMenu(Guid oGuid) : base(oGuid)
        {
            Caption = new DTOLangText(base.Guid, DTOLangText.Srcs.WebMenuItem);
            Rols = new List<DTORol>();
            Items = new List<DTOMenu>();
        }

        public static DTOMenu Factory(DTOLangText caption, string url = "")
        {
            DTOMenu retval = new DTOMenu();
            retval.Caption = caption;
            retval.LangUrl = new DTOLangText(url);
            return retval;
        }

        public static DTOMenu Factory(string nom, string url = "") // TO DEPRECATE
        {
            DTOMenu retval = new DTOMenu();
            {
                var withBlock = retval;
                withBlock.Caption.Esp = nom;
                withBlock.LangUrl = new DTOLangText(url);
            }
            return retval;
        }

        public static DTOMenu Factory(DTOLangText nom, DTOLangText url)
        {
            DTOMenu retval = new DTOMenu();
            retval.Caption = nom;
            retval.LangUrl = url;
            return retval;
        }

        public DTOMenu AddChild(DTOLangText nom, string url)
        {
            var retval = DTOMenu.Factory(nom, url);
            Items.Add(retval);
            return retval;
        }

        public DTOMenu AddChild(DTOLangText nom, DTOLangText url)
        {
            var retval = DTOMenu.Factory(nom, url);
            Items.Add(retval);
            return retval;
        }


        //public static string SegmentUrl(DTOMenu oMenu, params string[] oParams)
        //{
        //    System.Text.StringBuilder sb = new System.Text.StringBuilder();
        //    sb.Append(oMenu.Url);
        //    foreach (string sParam in oParams)
        //        sb.Append("/" + sParam);
        //    string retval = sb.ToString();
        //    return retval;
        //}

        public override string ToString()
        {
            return string.Format("DTOMenu: {0}", Caption.Esp);
        }

        public class Collection : List<DTOMenu>
        {
            public DTOMenu Add(DTOLangText nom, DTOLangText url)
            {
                var retval = DTOMenu.Factory(nom, url);
                base.Add(retval);
                return retval;
            }
        }
    }

}

File: matiasmasso/Mat.NET/DT2/DTOMenu.vb
Public Class DTOMenu
    Inherits DTOBaseGuid

    Property App As DTOApp.AppTypes
    Property Cod As Cods

    Property Caption As DTOLangText
    Property Url As String
    Property Ord As String
    Property Action As String
    Property Controller As String
    Property Rols As List(Of DTORol)
    Property Parent As DTOMenu
    Property Items As List(Of DTOMenu)

    Public Enum Cods
        NotSet
        Queries
        Forms
        Profile
        Customer
        Rep
        Comercial
        Staff
    End Enum

    Public Sub New()
        MyBase.New()
        _Caption = New DTOLangText(MyBase.Guid, DTOLangText.Srcs.WebMenuItem)
        _Rols = New List(Of DTORol)
        _Items = New List(Of DTOMenu)
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
        _Caption = New DTOLangText(MyBase.Guid, DTOLangText.Srcs.WebMenuItem)
        _Rols = New List(Of DTORol)
        _Items = New List(Of DTOMenu)
    End Sub

    Shared Function Factory(nom As String, Optional url As String = "") As DTOMenu
        Dim retval As New DTOMenu
        With retval
            .Caption.Esp = nom
            .Url = url
        End With
        Return retval
    End Function

    Public Function AddChild(nom As String, url As String) As DTOMenu
        Dim retval = DTOMenu.Factory(nom, url)
        _Items.Add(retval)
        Return retval
    End Function


    Shared Function SegmentUrl(oMenu As DTOMenu, ParamArray oParams() As String) As String
        Dim sb As New System.Text.StringBuilder
        sb.Append(oMenu.Url)
        For Each sParam As String In oParams
            sb.Append("/" & sParam)
        Next
        Dim retval As String = sb.ToString
        Return retval
    End Function

    Public Class Collection
        Inherits List(Of DTOMenu)
    End Class
End Class


